instrumentation/opentelemetry-instrumentation-pymongo/src/opentelemetry/instrumentation/pymongo/__init__.py

```diff
@@ -48,14 +85,34 @@
 from opentelemetry.instrumentation.utils import _SUPPRESS_INSTRUMENTATION_KEY
 from opentelemetry.semconv.trace import DbSystemValues, SpanAttributes
 from opentelemetry.trace import SpanKind, get_tracer
+from opentelemetry.trace.span import Span
 from opentelemetry.trace.status import Status, StatusCode
 
+_LOG = getLogger(__name__)
+
+RequestHookT = Callable[[Span, monitoring.CommandStartedEvent], None]
+ResponseHookT = Callable[[Span, monitoring.CommandSucceededEvent], None]
+FailedHookT = Callable[[Span, monitoring.CommandFailedEvent], None]
+
+
+def dummy_callback(span, event):
+    ...
+
 
 class CommandTracer(monitoring.CommandListener):
-    def __init__(self, tracer):
+    def __init__(
+        self,
+        tracer,
+        request_hook: RequestHookT = dummy_callback,
+        response_hook: ResponseHookT = dummy_callback,
+        failed_hook: FailedHookT = dummy_callback,
+    ):
         self._tracer = tracer
         self._span_dict = {}
         self.is_enabled = True
+        self.start_hook = request_hook
+        self.success_hook = response_hook
```

Comment:
Why do we need separate hooks for success and failure? Can't we use the same one and simply send `None` as the error on success?

Reply:
the pymongo instrumentor invokes two different callbacks in case of [success](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-pymongo/src/opentelemetry/instrumentation/pymongo/__init__.py#L97) or [failure](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-pymongo/src/opentelemetry/instrumentation/pymongo/__init__.py#L108) 

each of this callbacks receive a different type of event as an argument. technically we can use one callback for both cases and check the type of the event in it. 

but I think it can be simpler with the addition of the failed_hook, in our case for example we won't pass this hook at all, since we don't collect extra data on failure, so we can just pass a simple response hook and assume the event is always a success event

Reply:
Ok, again let's see what the maintainers say.

Also, we will want to collect the error in case of failure, why not?

Reply:
it is already done by the [instrumentation](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-pymongo/src/opentelemetry/instrumentation/pymongo/__init__.py#L118) 
We will only use the hooks for collecting the response from mongo, everything else is already recorded on the span, but I'm adding the request_hook and the failure_hook anyway for future use and for others to use

Reply:
👍